### PR TITLE
[lock/unlock] OP stack scripts

### PR DESCRIPTION
**Pull request by makramkd**
## Motivation

We need to be able to prove optimism withdrawals and finalize them in Go.

## Solution

Implement the withdrawal proving logic in Go. Most of this PR is generated code and code from the Optimism contracts in order to generate the needed gethwrappers and mocks.

Reviewers should focus on the following files:

* core/services/ocr2/plugins/rebalancer/bridge/opstack/withdrawprover/prover.go: contains the core proving logic and it's structure follows the optimism SDK for the most part.
* core/services/ocr2/plugins/rebalancer/bridge/opstack/rlphelpers/rlp.go: contains needed RLP decoding capabilities that are not available out of the box from go-ethereum.

### CODE REVIEW

File: contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL1BridgeAdapter.sol [hunk 1]
```diff
@@ -46,17 +46,22 @@ contract OptimismL1BridgeAdapter is IBridgeAdapter {
       revert MsgShouldNotContainValue(msg.value);
     }
 
+    // Extra data for the L2 deposit.
+    // We encode the nonce in the extra data so that we can track the L2 deposit offchain.
+    bytes memory extraData = abi.encode(s_nonce++);
```
Comment (makramkd) [inline contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL1BridgeAdapter.sol:51]: Can revisit this later.

File: contracts/src/v0.8/rebalancer/interfaces/optimism/IOptimismCrossDomainMessenger.sol [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// SPDX-License-Identifier: BUSL-1.1
```
Comment (makramkd) [inline contracts/src/v0.8/rebalancer/interfaces/optimism/IOptimismCrossDomainMessenger.sol]: Change license to MIT

File: contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL2BridgeAdapter.sol
Comment (makramkd) [inline contracts/src/v0.8/rebalancer/bridge-adapters/OptimismL2BridgeAdapter.sol:1]: Add doc comment on OP bridge operation

File: core/scripts/ccip/rebalancer/opstack/prove_withdrawal.go [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+}
+
+func getMessageBedrockOutput(
+	optimismPortalAddress,
+	l2OutputOracleAddress common.Address,
+	l2BlockNumber *big.Int,
+	l1Client *ethclient.Client,
+) bedrockOutput {
+	if getFPAC(optimismPortalAddress, l1Client) {
+		panic("unimplemented")
```
Comment (makramkd) [inline core/scripts/ccip/rebalancer/opstack/prove_withdrawal.go]: Still needs to be implemented in order to support fault proof upgrade, see https://oplabs.notion.site/Fault-Proof-Mainnet-Partner-Upgrade-Mission-Control-1ebc8767955044a9a3c31d11647ddd0d

File: core/scripts/ccip/rebalancer/opstack/prove_withdrawal.go [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+
+// func maybeAddProofNode(
+// 	key [32]byte,
+// 	proof []hexutil.Bytes,
+// ) []hexutil.Bytes {
+// 	modifiedProof := make([]hexutil.Bytes, len(proof))
+// 	copy(modifiedProof, proof)
+// 	finalProofEl := modifiedProof[len(modifiedProof)-1]
+// 	finalProofElDecoded := rlp.DecodeBytes(finalProofEl)
+// }
```
Comment (makramkd) [inline core/scripts/ccip/rebalancer/opstack/prove_withdrawal.go]: This is done in the SDK but not sure if this needs to be done. Unfortunately RLP decoding of Merkle Trie nodes is not straightforward in Go, will have to figure something out.

**Comment by makramkd** [comment]
Contracts are ready for review @matYang

File: core/scripts/ccip/rebalancer/opstack/prove_withdrawal.go (hunk 4 above)
Comment (makramkd) [reply core/scripts/ccip/rebalancer/opstack/prove_withdrawal.go]: Implemented now.

File: contracts/src/v0.8/rebalancer/interfaces/optimism/DisputeTypes.sol
Comment (makramkd) [inline contracts/src/v0.8/rebalancer/interfaces/optimism/DisputeTypes.sol:1]: Add source link and commit hash for all these files.

**Review by dimkouv — APPROVED** [review]
great work!

**Review by matYang — APPROVED** [review]

**Pull request merged**